haystack/components/converters/output_adapter.py

```diff
@@ -0,0 +1,130 @@
+import json
+from typing import Optional, Dict, Any, Set, Callable
+
+import jinja2.runtime
+from jinja2 import TemplateSyntaxError, meta
+from jinja2.nativetypes import NativeEnvironment
+from haystack.utils.type_serialization import serialize_type, deserialize_type
+from haystack import component, default_to_dict, default_from_dict
+
+
+class OutputAdaptationException(Exception):
+    """Exception raised when there is an error during output adaptation."""
+
+
+@component
+class OutputAdapter:
+    """
+    OutputAdapter in Haystack 2.x pipelines is designed to adapt the output of one component
+    to be compatible with the input of another component using Jinja2 template expressions.
+
+    The component configuration requires specifying the adaptation rules. Each rule comprises:
+    - 'template': A Jinja2 template string that defines how to adapt the input data.
+    - 'output_type': The type of the output data (e.g., str, List[int]).
+    - 'custom_filters': A dictionary of custom Jinja2 filters to be used in the template.
+
+    Example configuration:
+
+    ```python
+    from haystack.components.converters import OutputAdapter
+    adapter = OutputAdapter(template="{{ documents[0].content }}", output_type=str)
+
+    input_data = {"documents": [{"content": "Test content"}]}
+    expected_output = {"output": "Test content"}
+
+    assert adapter.run(**input_data) == expected_output
+    ```
+
+    In the pipeline setup, the adapter is placed between components that require output/input adaptation.
+    The name under which the adapted value is published is `output`. Use this name to connect the OutputAdapter
+    to downstream components in the pipeline.
+
+    Example pipeline setup:
+
+    """
+
+    predefined_filters = {
+        "json_loads": lambda s: json.loads(s) if isinstance(s, str) else json.loads(str(s)),
```

Comment:
On paper this is a security concern, loading arbitrary JSON data might be dangerous (it's not by chance that Jinja doesn't have it among its predefined filters). Let's discuss risks before merging.
```suggestion
        "fromjson": json.loads,
```

Reply:
Can we offload the burden of responsibility to users by not having any "risky" preinstalled filters? If users add filters, that's on them. It's similar to having the ability to add filters to the jinja2 NativeEnvironment. We shift the responsibility to the user.